### PR TITLE
3.7win without IPC fix

### PR DESCRIPTION
**Pull request by bagong**
Consider this a discussion base. It basically contains most from the old Winport branch except Lucas Cornelisses IPC fix. So it's a Win version that builds but has the bug.

I added:
- Miguel's open long files PR.
- tims commit containing most of Lucas' stuff for VS64bit compatibility (I "cherry-pick"ed that single commit from master. Not sure what else that brought in that wasn't planned to be in the 3.7 branch)
- other smaller stuff from Lucas' port that allow the VS build to succeed (like an alternative to sys/time.h and some small Qt-related stuff
- cmake changes/additions that make building with VS and msys simple
- a README that is in sync with those cmake changes

99% of this is purely Windows related and doesn't effect the other OSes.

In theory somebody could try to take Lucas' IPC fix and apply it on a branch that branches from this release branch. That would allow for a Windows version that is still fairly close to 3.7, and the other OSes wouldn't be affected by the IPC fix... Unfortunately I cannot do this myself, but merging this would provide a useful base...

### CODE REVIEW

**Comment by crucialfelix** [comment]
Awesome !

File: CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,19 @@ if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
 	message(WARNING "WARNING: IN-PLACE BUILDS ARE NOT RECOMMENDED - PLEASE USE A BUILD DIRECTORY")
 endif()
 
+# allows cmake-find modules and Windows installer to identify the right architecture
+if(WIN32)
+    if(CMAKE_GENERATOR MATCHES "64" OR "$ENV{MSYSTEM}" MATCHES "64")
```
Comment (timblechmann) [inline CMakeLists.txt]: `CMAKE_CL_64` to detect 64bit msvc.

File: CMakeLists.txt [hunk 2]
```diff
@@ -244,9 +261,10 @@ if (CMAKE_COMPILER_IS_INTEL AND NOT WIN32)
 	add_definitions(-diag-disable 279) # controlling expression is constant
 endif()
 
-if(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
+if(x${CMAKE_CXX_COMPILER_ID} STREQUAL "xMSVC")
```
Comment (timblechmann) [inline CMakeLists.txt]: `if( CMAKE_CXX_COMPILER_ID  STREQUAL  MSVC )`

Comment (timblechmann) [inline CMakeLists.txt]: or:
https://cmake.org/cmake/help/v3.4/variable/MSVC.html

File: CMakeLists.txt [hunk 3]
```diff
@@ -365,7 +383,10 @@ else()
 	)
 endif()
 
-if(NOT APPLE)
+if(CMAKE_GENERATOR MATCHES "Visual Studio")
+	install(DIRECTORY sounds   CONFIGURATIONS Debug;Release DESTINATION "${auxresourcesdir}")
```
Comment (timblechmann) [inline CMakeLists.txt]: remove `Debug;Release` configurations: there is more than that

File: CMakeLists.txt [hunk 4]
```diff
@@ -398,7 +419,29 @@ else()
...
+	endif()
+
+	install(FILES "${QT_PLATFORM_PLUGIN_PREFIX}/qwindows.dll"
+		CONFIGURATIONS Release
+		DESTINATION "${CMAKE_INSTALL_PREFIX}/${auxresourcesdir}/platforms"
+	)
+	install(FILES "${QT_PLATFORM_PLUGIN_PREFIX}/qwindowsd.dll"
+		CONFIGURATIONS Debug
+		DESTINATION "${CMAKE_INSTALL_PREFIX}/${auxresourcesdir}/platforms"
+	)
```
Comment (timblechmann) [inline CMakeLists.txt]: this should go into `editors/scide/CMakeLists.txt`, to the same place where the platform plugins for cocoa are deployed.
adding a post-build step would also make the non-installed binaries runnable

File: HelpSource/Guides/Working_with_HID.schelp [hunk 5]
```diff
@@ -1,6 +1,6 @@
 title:: Working with HID
 summary:: A guide to using HID devices for control input
-categories:: External control>HID
+categories:: External Control>HID
```
Comment (timblechmann) [inline HelpSource/Guides/Working_with_HID.schelp]: could these typos be moved into another PR? it is not exactly related ....

File: QtCollider/primitives/prim_misc.cpp [hunk 6]
```diff
@@ -167,7 +167,7 @@ QC_LANG_PRIMITIVE( Qt_SetGlobalPalette, 1, PyrSlot *r, PyrSlot *a, VMGlobals *g
 {
   if( !QcApplication::compareThread() ) return QtCollider::wrongThreadError();
 
-  QPalette p = QtCollider::get( a );
+  QPalette p = (QPalette&&) QtCollider::get(a);
```
Comment (timblechmann) [inline QtCollider/primitives/prim_misc.cpp]: casting to an rvalue reference??? this looks totally wrong.

File: editors/sc-ide/CMakeLists.txt [hunk 7]
```diff
@@ -350,21 +350,29 @@ if(APPLE)
                 #fixup_qt5_executable(\"${CMAKE_INSTALL_PREFIX}/SuperCollider/SuperCollider.app\")
                 " COMPONENT Runtime)
 
-
+elseif(CMAKE_GENERATOR MATCHES "Visual Studio")
+  set_property(TARGET SuperCollider PROPERTY RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/editors/sc-ide")
+  install(TARGETS SuperCollider
+      CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: `CONFIGURATIONS Debug;Release` remove.

File: editors/sc-ide/CMakeLists.txt [hunk 8]
```diff
@@ -350,21 +350,29 @@ if(APPLE)
...
+      DESTINATION "SuperCollider"
+      PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
+  install(FILES ${translations}
+      CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: `CONFIGURATIONS Debug;Release` remove.

... probably there are more instances, i won't comment on all of them

File: editors/sc-ide/widgets/code_editor/autocompleter.cpp [hunk 9]
```diff
@@ -807,9 +807,9 @@ void AutoCompleter::updateCompletionMenuInfo()
     QString examples = parseClassElement(node, "EXAMPLES");
     if (!examples.isEmpty())
         examples.prepend("<h4>Examples</h4>");
-
-    QString infos = QStringLiteral("<h4>%1</h4>%2%3"
-                            "<p><a href=\"%4\">go to help</a>")
+// MSVStudio 2013 does not concatenate multiple QStringliterals ("""") properly
+// see blog.qt.io/blog/2014/06/13/qt/weekly/13/qstringliteral/
```
Comment (timblechmann) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: comments are not helpful, if they point to a non-existing website

File: lang/LangSource/Bison/lang11d [hunk 10]
```diff
@@ -46,20 +46,20 @@ root	: classes
 
 classes : { $$ = 0; }
 		|	classes classdef
-			{ $$ = (long)linkNextNode((PyrParseNode*)$1, (PyrParseNode*)$2); }
+			{ $$ = (intptr_t)linkNextNode((PyrParseNode*)$1, (PyrParseNode*)$2); }
```
Comment (timblechmann) [inline lang/LangSource/Bison/lang11d]: this is already in:
c0a550d247ef98088dba3ca12c0e49587c2b6748

File: server/plugins/CMakeLists.txt [hunk 11]
```diff
@@ -255,9 +255,18 @@ elseif (APPLE)
 endif()
 
 if (WIN32)
-	install(TARGETS ${plugins} ${supernova_plugins}
-			DESTINATION "SuperCollider/plugins"
-			PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE)
+    if(CMAKE_GENERATOR MATCHES "Visual Studio")
+        set_property(TARGET  ${plugins} ${supernova_plugins} PROPERTY LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/server/plugins")
+        install(TARGETS ${plugins} ${supernova_plugins}
+            CONFIGURATIONS Debug;Release
```
Comment (timblechmann) [inline server/plugins/CMakeLists.txt]: .

File: server/scsynth/CMakeLists.txt [hunk 12]
```diff
@@ -187,7 +187,9 @@ else()
 endif()
 
 if (WIN32)
-	target_link_libraries(libscsynth wsock32 ws2_32 winmm)
+	target_link_libraries(libscsynth wsock32 ws2_32 winmm kernel32 user32
+		gdi32 winspool shell32 ole32 oleaut32 uuid comdlg32 advapi32 tlsf
+		boost_thread boost_system)
```
Comment (timblechmann) [inline server/scsynth/CMakeLists.txt]: tlsf boost_thread boost_system ... all of them are only used on windows???

**Comment by timblechmann** [comment]
thanks for picking this up. i've reviewed this again ... and apart from the small comments:

* please try to rebase it: it contains some commits, which are already in master
* in general: the smaller and more precise the change set is, the easier it is to review. better to more small PRs than one huge.

**Comment by bagong** [comment]
Thanks a lot for the review! I will go through it during the day. The commits from master came in, when I cherry-picked your commit that brings in Lucas' 64-bit related stuff. I hope I can find a good way to remove the master elements in it.

File: lang/LangSource/Bison/lang11d (hunk 10 above)
Comment (bagong) [inline lang/LangSource/Bison/lang11d]: Well, I cherry picked c0a550d to the 3.7 branch, it is only on master. That is also how the other stuff from master came in.

File: editors/sc-ide/CMakeLists.txt (hunk 8 above)
Comment (bagong) [inline editors/sc-ide/CMakeLists.txt]: The problem here is the following: with the multi configuration VS generator the "build configuration" or "cmake build type" is not determined in the the cmake configure step, but as an argument to "make". "Make" expects alternative configurations to be available for the build. In the case of portaudio, you get errors if you use a static library compiled as release version in a SC debug build and vice versa. So rather than just switch the configuration, you'd have to point to different port-audio source as well, which is a bit hassle. This change allows to access differently configured portaudio libraries. I know there are more configurations, but - at least at the time I did this - cmake only allowed to handle two configurations in this manner. Do you know an alternative way to handle the underlying problem?
On a side line: Yesterday @sonoro1234 published a portaudio version that builds successfully with cmake using the MinGW compiler, and can also be used to compile with VS (https://github.com/sonoro1234/portaudio20111121) - the problem I tried to solve with above changes would go away if portaudio was added to sc as external module. Then portaudio will just be compiled in the configuration chosen for the SC build. I know you actually did that in the past, but then removed it again. What is your take on adding portaudio to the SC sources? Why did you remove it in the past?

Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: as for portaudio, there are two options:

link it as dll and don't care about window/static linking issues (matching compiler options, crt and the like)

use an imported library, like:
```
add_library(portaudio STATIC IMPORTED GLOBAL)

set_target_properties(portaudio PROPERTIES
  IMPORTED_LOCATION_DEBUG   /path/to/portaudio_debug.lib
  IMPORTED_LOCATION                  /path/to/portaudio_release.lib
)
```

portaudio-as-dll might be the most robust way. embeding portaudio into the sc build process shouldn't be too difficult, either, someone just needs to do it.

Comment (bagong) [inline editors/sc-ide/CMakeLists.txt]: > someone just needs to do it
Yea, or more precise, somebody who knows how to do it needs to do it. Using a dll would force us to provide a binary, as portaudio is not distributed as binary, or expect people to build their own dll. Building a portaudio dll is even more fragile than a static library, I haven't digged deep into @sonoro1234 cmake mods, but I did see errors when building more than just the static library. (Check Google, I am not the only one with this problem. The build configuration pulished here (https://github.com/slowriot/portaudio-mingw-w64) breaks when ASIO support is added).
My conclusion tends to be, embed portaudio and have a static library being built. Can you or somebody else give pointers of how to do that?

File: QtCollider/primitives/prim_misc.cpp (hunk 6 above)
Comment (bagong) [inline QtCollider/primitives/prim_misc.cpp]: This is the error produced by VS with the original:
  C:\Users\Rainer\Projects\sc\supercollider\QtCollider\primitives\prim_misc.cpp(170): error C2440: 'initializing' : can
not convert from 'QtCollider::DecodableSlot' to 'QPalette' [C:\Users\Rainer\Projects\sc\supercollider\build3.7VS64\lang
\libsclang.vcxproj]

File: editors/sc-ide/CMakeLists.txt (hunk 8 above)
Comment (timblechmann) [inline editors/sc-ide/CMakeLists.txt]: embedding portaudio is trivial, collecting all dependencies is not so trivial ... iirc kernel streaming and/or WASAPI require certain parts of the windows sdk to be installed manually, asio requires the asio sdk to be installed manually. the portaudio-part is simpler, though

File: QtCollider/primitives/prim_misc.cpp (hunk 6 above)
Comment (timblechmann) [inline QtCollider/primitives/prim_misc.cpp]: with a raw c-style cast anything can be made to compile, in this case it looks super-dangerous.

maybe something like this works?
`static_cast<QPalette>( QtCollider::get( a ) );`

File: QtCollider/style/ProxyStyle.cpp [hunk 13]
```diff
@@ -19,8 +19,9 @@ static bool AlwaysShowScrollbars() {
 #elif Q_OS_X11
   return !QcApplication::SystemHasMouseWheel();
 
-#elif Q_OS_WIN
-  return !QcApplication::SystemHasMouseWheel();
+// Creates a build error in VStudio 2013 C1017
+// #elif Q_OS_WIN
+//  return !QcApplication::SystemHasMouseWheel();
```
Comment (timblechmann) [inline QtCollider/style/ProxyStyle.cpp]: these should all be ifdefs or `if defined(Q_OS_XXX)`

File: editors/sc-ide/CMakeLists.txt (hunk 8 above)
Comment (bagong) [inline editors/sc-ide/CMakeLists.txt]: wasapi, wdmks and dsound are not supported in the old portaudio that shipped with SC, and all three in fact hard to compile, especially in an MinGW environment. The current VS build I was using had all those apis built, but I was wondering if anybody ever used them. One price of the versatility is that you get an enormous list of devices when the server boosts, which is probably more confusing than helpful, and of course the build get's more fragile.
Do you agree that WMME and Asio are enough? In that case, building with ASIO or not would only depend on whether the person who builds SC adds the ASIO SDK in a sibling folder of the portaudio source (and dsound would have to be hardcoded to OFF to guard against building it on machines where directX is installed (80% of the attempts to build with DX support fail - only certain pa-dx-compiler-combinations work)).

File: QtCollider/primitives/prim_misc.cpp (hunk 6 above)
Comment (bagong) [inline QtCollider/primitives/prim_misc.cpp]: Unfortunately not:
C:\Users\Rainer\Projects\sc\supercollider\QtCollider\primitives\prim_misc.cpp(174): error C2440: 'static_cast' : cannot convert from 'QtCollider::DecodableSlot' to 'QPalette' [C:\Users\Rainer\Projects\sc\supercollider\build3.7VS64\lang\libsclang.vcxproj]

**Comment by bagong** [comment]
This is the error that triggered the first build to fail on Linux (I regriggered):

travis_time:end:2ba0b8a8:start=1455023702481277606,finish=1455024319003218018,duration=616521940412
[0K
[32;1mThe command "make install" exited with 0.[0m
travis_time:start:0bc9fd4b
[0K$ $TRAVIS_BUILD_DIR/testsuite/sclang/launch_test.py $SCLANG
Traceback (most recent call last):
  File "/home/travis/build/supercollider/supercollider/testsuite/sclang/launch_test.py", line 31, in <module>
    assert os.path.exists(sclang_path)
AssertionError

travis_time:end:0bc9fd4b:start=1455024319008010345,finish=1455024319082033013,duration=74022668
[0K
[31;1mThe command "$TRAVIS_BUILD_DIR/testsuite/sclang/launch_test.py $SCLANG" exited with 1.[0m
travis_time:start:081a772d
[0K$ export COMMIT_NAME=$TRAVIS_COMMIT

travis_time:end:081a772d:start=1455024319086130758,finish=1455024319089472629,duration=3341871
[0K
[32;1mThe command "export COMMIT_NAME=$TRAVIS_COMMIT" exited with 0.[0m
travis_time:start:240b0b48
[0K$ ifmac mkdir -p $HOME/artifacts

travis_time:end:240b0b48:start=1455024319093354174,finish=1455024319096417490,duration=3063316
[0K
[32;1mThe command "ifmac mkdir -p $HOME/artifacts" exited with 0.[0m
travis_time:start:09e8646c
[0K$ (ifmac cd Install; ifmac zip -r $HOME/artifacts/SC-$COMMIT_NAME.zip SuperCollider)

travis_time:end:09e8646c:start=1455024319100602544,finish=1455024319104399347,duration=3796803
[0K
[32;1mThe command "(ifmac cd Install; ifmac zip -r $HOME/artifacts/SC-$COMMIT_NAME.zip SuperCollider)" exited with 0.[0m

Done. Your build exited with 1.

**Comment by bagong** [comment]
So I checked on that build error: 
- I can build this PR just fine on Ubuntu-Studio 15.10
- I do get that python error there as well
- I get *the same* python-error if I build the 3.7 vanilla branch on Ubuntu

So why does it create an error here, and not in the other builds?

**Comment by bagong** [comment]
The Travis error is the same one as before...

**Comment by bagong** [comment]
For some reason Travis hasn't commited that las commit. But according to Winmerge this branch is now identical with the testbranch I was building up... And there it builds..

**Comment by bagong** [comment]
I am done with this. It looks as Travis didn't build the last commit, but as far as I can see, it did, only it seems to be out of sync with my commits. The ad8d9dc (second last) commit is marked okay now, although that was marked failed when it was initially triggered. I did try to retrigger the pull-request manually and it still. What I think happened is that it compiled the last commit but marked the second last commit. The last commit, according to winmerge is exactly identical with the last commit on the testTravis fake PR I made to see when the error actually comes up. It seems to be the error is triggered when in cmake path expressions that contain variables are not wrapped in string delimiters. Not in every case, but in the ones I corrected. That python test is completely intransparent...
So I think I met all request by Tim 
- except the "feed my grandma cast problem", which I put in OS check if's to use it only when we compile with VS: https://github.com/bagong/supercollider/commit/39a766aec3f5b4e3a5cf7eb6ec6c420e98c41318
- The Q_OS_XXX thing required to reformulate the ifdefs as suggested by Tim. Apparently that's a Qt thing, I saw it somewhere on stackoverflow. I also include QtGlobal. Including QtGlobal only (without the syntactical adjustment) actually made the build break for MinGW too at this place, so it may be that this construction was simply ignored on some platforms.
- The biggest thing is the portaudio integration. Note that this is ignored on other platforms that Windows. I tested on both Linux and Apple that their build system is not contaminated when building with portaudio. My consideration for the inclusion of PA: on Windows you cannot get binaries of PA. Furthermore it is really difficult to compile PA using cmake if you want to get anything more than the basic WMME support. The only way to build PA fairly easily on Windows with other sound-api's is using msys and the unix build tools. That again is a different computer culture which I now tried to make unnecessary for building SC on Windows. What happens with this setup is: if a user does not provide PS libraries, SC will build a basic version as part of the SC build. If he/she does provide a library, that library will be preferred (can static or dynamic). Something I didn't manage is to allow the in place/embedded build of pa compile anything else than WMME support, but: the pa source is patched in a way that if he or she build PA separately from the embedded pa, the cmake build is going to succeed for all audio APS's except WASAPI. I puzzled together patches from 3 different sources until it succeeded. You can see the patches in the submodule, it's a git repo and the patches are commits. I am happy to move that git repo to some neutral place. It git svn's the portaudio svn repo and is at latest pa-svn
- remains the master commits that came in with the big Tim commit which I cherry-picked from master. In my opinion we shouldn't be too formalistic: these are fixes of typos in documentation, right. What bad can come from them in the 3.7 branch? So the only relevant one is Dan's https://github.com/bagong/supercollider/commit/a630e05b67cd60d801af87db1373428f4da832ee. Should I revert it, or is it actually good to have it in 3.7?

best
.r.

**Comment by bagong** [comment]
Lucas' IPC fix is also done. It's a single commit on top of this, nicely separated. So with that commit Windows is stable, remains MIDI and the interactive console from the obvious things...

**Comment by bagong** [comment]
Great! ;)

**Comment by crucialfelix** [comment]
ok, now done ? (you added a few commits yesterday)

As you say, the big thing is adding portaudio in. The submodule will be there even for non-windows builds. This is a downside. Its your own fork, and that may causes stress/maintenance issues later. But at least we have windows running now, and the alternatives are worse. It can be easily switched for another (possibly an pa official) version in the future.

That read me file is pretty epic. I didn't read it all.  In the end do Anna Krenina and Count Vronsky live happily ever after ? 

We should merge this now, oder ?

You are a hero @bagong !

**Comment by bagong** [comment]
Thanks! Yep, merge ;)
As to portaudio: as I stated, I am happy to move the PA submodule to the supercollider repo, I would actually prefer that...

**Comment by crucialfelix** [comment]
If there is any need to change it or burden on you then we can move it into supercollider when that time comes.

**Comment by bagong** [comment]
okle

**Comment by bagong** [comment]
I have pushed Lucas' "IPCfix" that makes the Windows version stable as branch "topic/IPCwithQTcp. Depending on what happens until the release, it could be used to make the win release stable... I'll try to keep it on top of 3.7...
